# Data Analysis with NumPy & Pandas
## Exploring the Iris Dataset

This notebook demonstrates exploratory data analysis and basic statistics on the classic Iris dataset using Python's `pandas` and `numpy` libraries.


In [1]:
import numpy as np
import pandas as pd
from scipy import stats

# Load the Iris dataset directly from a public URL
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/iris.csv"
data = pd.read_csv(url)

print("Dataset loaded successfully!")
print(f"Shape: {data.shape[0]} rows × {data.shape[1]} columns")


Dataset loaded successfully!
Shape: 150 rows × 5 columns


---
## 1. Explore the Data


In [2]:
# ── First few rows ──────────────────────────────────────────────────────────
print("First 5 rows of the dataset:")
data.head()


First 5 rows of the dataset:


,sepal_length,sepal_width,petal_length,petal_width,species
0,5.1,3.5,1.4,0.2,setosa
1,4.9,3.0,1.4,0.2,setosa
2,4.7,3.2,1.3,0.2,setosa
3,4.6,3.1,1.5,0.2,setosa
4,5.0,3.6,1.4,0.2,setosa


In [3]:
# ── Data types and non-null counts ──────────────────────────────────────────
print("Dataset Info:")
data.info()


Dataset Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal_length  150 non-null    float64
 1   sepal_width   150 non-null    float64
 2   petal_length  150 non-null    float64
 3   petal_width   150 non-null    float64
 4   species       150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


In [4]:
# ── Summary statistics ───────────────────────────────────────────────────────
print("Summary Statistics:")
data.describe()


Summary Statistics:


,sepal_length,sepal_width,petal_length,petal_width
count,150.000000,150.000000,150.000000,150.000000
mean,5.843333,3.057333,3.758000,1.199333
std,0.828066,0.435866,1.765298,0.762238
min,4.300000,2.000000,1.000000,0.100000
25%,5.100000,2.800000,1.600000,0.300000
50%,5.800000,3.000000,4.350000,1.300000
75%,6.400000,3.300000,5.100000,1.800000
max,7.900000,4.400000,6.900000,2.500000


---
## 2. Calculate Basic Statistics


In [6]:
# ── Select only numeric columns ──────────────────────────────────────────────
numeric_cols = data.select_dtypes(include=[np.number])

# Mean
print("=" * 55)
print("MEAN")
print("=" * 55)
print(numeric_cols.mean().to_string())

# Median
print("\n" + "=" * 55)
print("MEDIAN")
print("=" * 55)
print(numeric_cols.median().to_string())

# Mode  (returns a DataFrame; take row 0 for the first mode)
print("\n" + "=" * 55)
print("MODE")
print("=" * 55)
print(numeric_cols.mode().iloc[0].to_string())

# Standard Deviation
print("\n" + "=" * 55)
print("STANDARD DEVIATION")
print("=" * 55)
print(numeric_cols.std().to_string())


MEAN
sepal_length    5.843333
sepal_width     3.057333
petal_length    3.758000
petal_width     1.199333

MEDIAN
sepal_length    5.80
sepal_width     3.00
petal_length    4.35
petal_width     1.30

MODE
sepal_length    5.0
sepal_width     3.0
petal_length    1.4
petal_width     0.2

STANDARD DEVIATION
sepal_length    0.828066
sepal_width     0.435866
petal_length    1.765298
petal_width     0.762238


In [7]:
# ── Combined statistics table ────────────────────────────────────────────────
stats_df = pd.DataFrame({
    "Mean":   numeric_cols.mean(),
    "Median": numeric_cols.median(),
    "Mode":   numeric_cols.mode().iloc[0],
    "Std Dev": numeric_cols.std(),
    "Min":    numeric_cols.min(),
    "Max":    numeric_cols.max(),
    "Range":  numeric_cols.max() - numeric_cols.min(),
})

print("Combined Statistics Table:")
stats_df.round(4)


Combined Statistics Table:


,Mean,Median,Mode,Std Dev,Min,Max,Range
sepal_length,5.8433,5.80,5.0,0.8281,4.3,7.9,3.6
sepal_width,3.0573,3.00,3.0,0.4359,2.0,4.4,2.4
petal_length,3.7580,4.35,1.4,1.7653,1.0,6.9,5.9
petal_width,1.1993,1.30,0.2,0.7622,0.1,2.5,2.4


In [8]:
# ── Correlation matrix ───────────────────────────────────────────────────────
print("Pearson Correlation Matrix:")
corr_matrix = numeric_cols.corr()
print(corr_matrix.to_string())
corr_matrix.round(4)


Pearson Correlation Matrix:
              sepal_length  sepal_width  petal_length  petal_width
sepal_length      1.000000    -0.117570      0.871754     0.817941
sepal_width      -0.117570     1.000000     -0.428440    -0.366126
petal_length      0.871754    -0.428440      1.000000     0.962865
petal_width       0.817941    -0.366126      0.962865     1.000000


,sepal_length,sepal_width,petal_length,petal_width
sepal_length,1.0000,-0.1176,0.8718,0.8179
sepal_width,-0.1176,1.0000,-0.4284,-0.3661
petal_length,0.8718,-0.4284,1.0000,0.9629
petal_width,0.8179,-0.3661,0.9629,1.0000


In [9]:
# ── Interpret the strongest correlations ────────────────────────────────────
print("Top Correlations (excluding self-correlation):\n")

# Flatten the upper triangle only
corr_pairs = (
    corr_matrix
    .where(np.triu(np.ones(corr_matrix.shape), k=1).astype(bool))
    .stack()
    .reset_index()
)
corr_pairs.columns = ["Feature A", "Feature B", "Correlation"]
corr_pairs["Abs Correlation"] = corr_pairs["Correlation"].abs()
corr_pairs = corr_pairs.sort_values("Abs Correlation", ascending=False).reset_index(drop=True)

print(corr_pairs[["Feature A", "Feature B", "Correlation"]].to_string(index=False))


Top Correlations (excluding self-correlation):

   Feature A    Feature B  Correlation
petal_length  petal_width     0.962865
sepal_length petal_length     0.871754
sepal_length  petal_width     0.817941
 sepal_width petal_length    -0.428440
 sepal_width  petal_width    -0.366126
sepal_length  sepal_width    -0.117570


In [10]:
# ── NumPy cross-check ────────────────────────────────────────────────────────
print("NumPy verification for 'sepal_length':")
col = data["sepal_length"].values

print(f"  np.mean   : {np.mean(col):.4f}")
print(f"  np.median : {np.median(col):.4f}")
print(f"  np.std    : {np.std(col, ddof=1):.4f}  (sample std, ddof=1)")
print(f"  np.min    : {np.min(col):.4f}")
print(f"  np.max    : {np.max(col):.4f}")
print(f"  np.var    : {np.var(col, ddof=1):.4f}  (sample variance)")


NumPy verification for 'sepal_length':
  np.mean   : 5.8433
  np.median : 5.8000
  np.std    : 0.8281  (sample std, ddof=1)
  np.min    : 4.3000
  np.max    : 7.9000
  np.var    : 0.6857  (sample variance)


---
## Summary

| Metric | What it tells us |
|--------|-----------------|
| **Mean** | The arithmetic average — sensitive to outliers |
| **Median** | The middle value — robust to outliers |
| **Mode** | The most frequently occurring value |
| **Std Dev** | Average spread of values around the mean |
| **Correlation** | Linear relationship strength between two features (–1 to +1) |

> **Key finding:** `petal_length` and `petal_width` show the strongest positive correlation (~0.96), meaning longer petals tend to be wider petals across the Iris species.
